src/v/datalake/proto_to_arrow_scalar.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include "datalake/proto_to_arrow_interface.h"
+
+#include <arrow/api.h>
+#include <arrow/array/array_primitive.h>
+#include <arrow/type.h>
+#include <google/protobuf/message.h>
+
+#include <memory>
+#include <stdexcept>
+
+namespace datalake::detail {
+
+template<typename ArrowType>
+class proto_to_arrow_scalar : public proto_to_arrow_interface {
+    using BuilderType = arrow::TypeTraits<ArrowType>::BuilderType;
+
+public:
+    proto_to_arrow_scalar()
+      : _builder(std::make_shared<BuilderType>()) {}
+
+    arrow::Status
+    add_value(const google::protobuf::Message* msg, int field_idx) override {
+        if (!_arrow_status.ok()) {
+            return _arrow_status;
+        }
+        do_add<ArrowType>(msg, field_idx);
+        return _arrow_status;
+    }
+
+    arrow::Status finish_batch() override {
+        if (!_arrow_status.ok()) {
+            return _arrow_status;
+        }
+        auto&& builder_result = _builder->Finish();
+        _arrow_status = builder_result.status();
+        std::shared_ptr<arrow::Array> array;
+        if (!_arrow_status.ok()) {
+            return _arrow_status;
+        }
+
+        // Safe because we validated the status after calling `Finish`
+        array = std::move(builder_result).ValueUnsafe();
+        _values.push_back(array);
+        return _arrow_status;
+    }
+
+    std::shared_ptr<arrow::Field> field(const std::string& name) override {
+        return arrow::field(
+          name, arrow::TypeTraits<ArrowType>::type_singleton());
+    }
+
+    std::shared_ptr<arrow::ArrayBuilder> builder() override { return _builder; }
+
+private:
+    template<typename T>
+    void
+    do_add(const google::protobuf::Message* /* msg */, int /* field_idx */) {
```

Comment:
Here and throughout- do you want `const google::protobuf::Message* const`? Is there a reason to prefer pointers to references?

Reply:
The Protobuf API takes returns pointers for objects and takes pointers as arguments. I try to avoid raw pointers as much as possible, but I get this impression this is the Protobuf way.

Reply:
Google generally uses raw pointers over l-value references in function parameters. This is because it forces the caller to annotate parameters as taking a pointer (via &).

Absl will have many of the same patterns (especially in the string methods like StripPrefix and SimpleAtoi)

Reply:
Also: https://github.com/redpanda-data/redpanda/pull/18245#discussion_r1597093063

Reply:
👋

Reply:
👋  don't mind me driving by!